### PR TITLE
Refactor core-backend errors

### LINKED ISSUE #1979: `WrongInstrumentation` programs shouldn't be able to be called again
Need to make decision on design if approve this

### PR DESCRIPTION
**Pull request by ark0f**
The second step to #1083, #1084

Resolves #1979, #1652

### CODE REVIEW

File: core-backend/common/src/lib.rs [hunk 1]
```diff
@@ -268,25 +154,23 @@ pub struct BackendReport<T, E> {
     pub ext: E,
 }
 
-#[derive(Debug, derive_more::Display)]
-pub enum StackEndError {
-    #[display(fmt = "Stack end addr {_0:#x} must be aligned to WASM page size")]
-    IsNotAligned(u32),
+#[derive(Debug)]
+pub enum EnvironmentExecutionError<B, P> {
```
Comment (gshep) [inline core-backend/common/src/lib.rs:158]: I personally prefer more longer names for generics

File: core-backend/common/src/mock.rs [hunk 2]
```diff
@@ -45,22 +45,24 @@ impl fmt::Display for Error {
...
     }
-}
 
-impl IntoExtError for Error {
     fn into_ext_error(self) -> Result<ExtError, Self> {
         todo!()
     }
+
+    fn into_termination_reason(self) -> TerminationReason {
+        todo!()
```
Comment (gshep) [inline core-backend/common/src/mock.rs:64]: maybe `unimplemented!` is better since we don't intend to implement it ever

**Review by gshep — APPROVED** [review]
Awesome work! 💯

File: core-processor/src/ext.rs [hunk 3]
```diff
@@ -894,7 +835,7 @@ impl EnvExt for Ext {
         let code_hash = packet.code_id();
 
         // Send a message for program creation
-        let result = self
+        let value = self
```
Comment (breathx) [inline core-processor/src/ext.rs:838]:
```suggestion
        let res = self
```

File: examples/binaries/syscall-error/src/lib.rs [hunk 4]
```diff
@@ -31,6 +31,21 @@ use gstd::errors::{ContractError, ExtError, MessageError};
 
 #[no_mangle]
 extern "C" fn init() {
+    unsafe {
+        let mut buf = [0; 1024];
+        let mut len = MaybeUninit::uninit();
```
Comment (breathx) [inline examples/binaries/syscall-error/src/lib.rs:36]: mv maybeuninint to buf, rm from len

File: gsys/src/lib.rs [hunk 5]
```diff
@@ -331,9 +331,9 @@ extern "C" {
     /// Fallible `gr_error` get syscall.
     ///
     /// Arguments type:
-    /// - `error`: `mut ptr` for buffer to store requested data.
-    /// - `err`: `mut ptr` for `u32` error length.
-    pub fn gr_error(error: *mut BufferStart, err: *mut Length);
+    /// - `buf`: `mut ptr` for buffer to store requested data.
+    /// - `len`: `mut ptr` for `u32` error length.
```
Comment (breathx) [inline gsys/src/lib.rs:335]:
```suggestion
    /// - `len`: `mut ptr` for `u32` current error length.
```

File: gsys/src/lib.rs [hunk 6]
```diff
@@ -331,9 +331,9 @@ extern "C" {
     /// Fallible `gr_error` get syscall.
     ///
     /// Arguments type:
-    /// - `error`: `mut ptr` for buffer to store requested data.
-    /// - `err`: `mut ptr` for `u32` error length.
-    pub fn gr_error(error: *mut BufferStart, err: *mut Length);
+    /// - `buf`: `mut ptr` for buffer to store requested data.
```
Comment (breathx) [inline gsys/src/lib.rs:334]:
```suggestion
    /// - `buf`: `mut ptr` for buffer to store previously occurred error.
```

File: pallets/gear/src/benchmarking/syscalls.rs [hunk 7]
```diff
@@ -1163,9 +1163,9 @@ where
 
     pub fn gr_error(r: u32) -> Result<Exec<T>, &'static str> {
```
Comment (breathx) [inline pallets/gear/src/benchmarking/syscalls.rs:1164]: Re-review bench values

**Review by breathx — CHANGES_REQUESTED** [review]

File: core-backend/sandbox/src/env.rs [hunk 8]
```diff
@@ -19,21 +19,16 @@
...
+    funcs::FuncsHandler as Funcs,
     memory::MemoryWrap,
     runtime::{self, Runtime},
 };
-use alloc::{
-    collections::BTreeSet,
-    string::{String, ToString},
-};
-use core::fmt;
+use alloc::{collections::BTreeSet, string::ToString};
```
[the remark this answers is not in the document]
Comment (ark0f) [reply core-backend/sandbox/src/env.rs:26]: Should be refactored though whole project. Area of work is not for this PR

File: core-backend/wasmi/src/funcs/internal.rs [hunk 9]
```diff
@@ -120,15 +124,15 @@ where
 
         f().ok_or_else(|| {
             // TODO #1979
```
[the remark this answers is not in the document]
Comment (ark0f) [reply core-backend/wasmi/src/funcs/internal.rs:126]: Will be done later

File: core-backend/common/src/lib.rs [hunk 10]
```diff
@@ -22,80 +22,55 @@
...
-}
-
-impl Deref for TrimmedString {
-    type Target = String;
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
+// '__gear_stack_end' export is inserted in wasm-proc or wasm-builder
```
Comment (shamilsan) [inline core-backend/common/src/lib.rs:65]:
```suggestion
// '__gear_stack_end' export is inserted by wasm-proc or wasm-builder
```

File: core-backend/common/src/lib.rs [hunk 11]
```diff
@@ -273,11 +273,17 @@ impl<Env, PrepMem> EnvironmentExecutionError<Env, PrepMem> {
     }
 }
 
-pub trait Environment<E, EP = DispatchKind>: Sized
+pub type EnvironmentExecutionResult<T, Env, EP> = Result<
```
Comment (breathx) [inline core-backend/common/src/lib.rs:277]: separate aliases

File: core-backend/sandbox/src/lib.rs [hunk 12]
```diff
@@ -21,6 +21,7 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
 extern crate alloc;
+extern crate core;
```
Comment (breathx) [inline core-backend/sandbox/src/lib.rs:24]: rm extern core

**Review by breathx — APPROVED** [review]

File: core-backend/common/src/lib.rs [hunk 13]
```diff
@@ -150,166 +125,269 @@ pub struct ExtInfo {
...
+        match self {
+            Ok(value) => Ok(Ok(value)),
+            Err(err) => {
+                *state_err = ActorSyscallFuncError::Core(err.clone()).into();
+                match err.into_ext_error() {
+                    Ok(ext_err) => Ok(Err(ext_err.encoded_size() as u32)),
+                    Err(err) => Err(ActorSyscallFuncError::Core(err)),
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (breathx) [reply core-backend/common/src/lib.rs:199]: Seems original code is more readable
No performance boost, profit in 4 lines doesn't cover cons

File: pallets/gear/src/benchmarking/syscalls.rs (hunk 7 above)
Comment (breathx) [reply pallets/gear/src/benchmarking/syscalls.rs:1164]: Discussed how to fix it

**Pull request merged**